File: Algorithmic Thinking, Models of Computation/peak_finding_1D.cpp
#include<iostream>
#include<bits/stdc++.h>
#include<ctime>
#include<cstdlib>
using namespace std;

int generate(vector<int> &A)
{
    int size = rand()%251;

    for(int i = 0; i < size; i++)
        A.push_back(rand()%1000);
    return size;
}

void print_vector(vector<int> &A)
{
    for(int x:A)
        cout<<x<<",";
    cout<<endl;
}

int peak(vector<int> &A, int lb, int ub)
{
    while(lb<=ub)
    {
      int mid = (lb+ub)/2;
      if(A[mid]<A[mid-1])
        ub = mid-1; 
      else if (A[mid]<A[mid+1])
        lb = mid+1;
      else
        return mid; 
    }
    
}

int main()
{
    srand(time(NULL));
    vector<int> A;
    int size = generate(A);
    cout<<size<<endl;
    print_vector(A);
    cout<<"-----------------------Result---------------------------\n";
    int pos = peak(A, 0, size-1);
    if (pos>0 && pos<size-1)
    cout<<"\n"<<"Position : "<<pos<<"\tValue : "<<A[pos]<<"\tVisualize area : ["<<A[pos-1]<<" "<<A[pos]<<" "<<A[pos+1]<<"]\n";
    else if (pos==0)
    cout<<"\n"<<"Position : "<<pos<<"\tValue : "<<A[pos]<<"\tVisualize area : ["<<A[pos]<<" "<<A[pos+1]<<"]\n";
    else if (pos == size-1)
    cout<<"\n"<<"Position : "<<pos<<"\tValue : "<<A[pos]<<"\tVisualize area : ["<<A[pos-1]<<" "<<A[pos]<<"]\n";
    cout<<"---------------------------------------------------------\n";

}

File: Algorithmic Thinking, Models of Computation/peak_finding_1D.py
from random import randint

def generate():
    size = randint(0, 10001)
    #size = 10000000
    l = [randint(0,100) for x in range(int(size))]
    return (int(size), l)

def solve(A, lb, ub):
    mid = (lb+ub)//2
    while(lb<ub):
        mid = (lb+ub)//2
        if A[mid]<A[mid-1]:
            ub = mid-1
        elif A[mid]<A[mid+1]:
            lb = mid+1
        else:
            return mid
    return mid

def main():
    size, A = generate()
    
    print(size)
    #print(A)
    pos = solve(A, 0, size - 1)
    print("--------------------------Result--------------------------")
    if pos>0 and pos<size-1:    
        print("Value : {}\tPosition : {}\tNeighbours : {}".format(A[pos], pos, A[pos-1:pos+2]))
    elif pos == 0:
        print("Value : {}\tPosition : {}\tNeighbours : {}".format(A[pos], pos, A[pos:pos+2]))
    elif pos == size - 1:
        print("Value : {}\tPosition : {}\tNeighbours : {}".format(A[pos], pos, A[pos-1:pos+1]))
    print("----------------------------------------------------------")

if __name__ == "__main__":
    main()